Hotfix/bug on stanford segmenter

xu2333 opened the pull request:
That’s exciting.

xu2333 committed:
The new version of stanford-segmenter-2016-10-31 don't need the path of slf4j

Now the stanford_segmenter.py is not working using the new release of
stanford-segmenter-2016-10-31.

Since the exception is threw when executing the java cmd, I tested the
following cmd and finally catch the bug.

1. work (the existing version of segment.sh from
stanford-segmenter-2016-10-31)

  java -mx2g -cp ./*: edu.stanford.nlp.ie.crf.CRFClassifier
-sighanCorporaDict ./data -textFile input.txt -inputEncoding UTF-8
-sighanPostProcessing true -keepAllWhitespaces false -loadClassifier
./data/ctb.gz -serDictionary ./data/dict-chris6.ser.gz

2. work (replace the . to absolute path from 1)

  java -mx2g -cp ./*: edu.stanford.nlp.ie.crf.CRFClassifier
-sighanCorporaDict
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data -textFile
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/input.txt -inputEncoding UTF-8 -sighanPostProcessing true
-keepAllWhitespaces false -loadClassifier
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/ctb.gz -serDictionary
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/dict-chris6.ser.gz

3. not work (replace ./* to specific files from 2)

  /usr/bin/java -mx2g -cp
//Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-1
0-31/stanford-segmenter-3.7.0.jar://Users/xu2333/Documents/4Paradigm/sta
nfordNLP/stanford-segmenter-2016-10-31/slf4j-1.7.25/slf4j-api-1.7.25.jar
: edu.stanford.nlp.ie.crf.CRFClassifier -sighanCorporaDict
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data -textFile
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/input.txt -inputEncoding UTF-8 -sighanPostProcessing true
-keepAllWhitespaces false -loadClassifier
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/ctb.gz -serDictionary
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/dict-chris6.ser.gz

4. not work (the existing version of stanford-segmenter.py)

  /usr/bin/java -mx2g -cp
//Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-1
0-31/stanford-segmenter-3.7.0.jar://Users/xu2333/Documents/4Paradigm/sta
nfordNLP/stanford-segmenter-2016-10-31/slf4j-1.7.25/slf4j-api-1.7.25.jar
 edu.stanford.nlp.ie.crf.CRFClassifier -loadClassifier
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/pku.gz -keepAllWhitespaces false -textFile input.txt
-serDictionary
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/dict-chris6.ser.gz -sighanCorporaDict
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/./data/ -sighanPostProcessing true -inputEncoding UTF-8

5. work (replace specific files to ./* from 4)

  /usr/bin/java -mx2g -cp ./*: edu.stanford.nlp.ie.crf.CRFClassifier
-loadClassifier
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/pku.gz -keepAllWhitespaces false -textFile input.txt
-serDictionary
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/dict-chris6.ser.gz -sighanCorporaDict
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/./data/ -sighanPostProcessing true -inputEncoding UTF-8

6. work (replace specific file names to * from 4) -> adapt as the new
version of stanford-segmenter.py

  /usr/bin/java -mx2g -cp
//Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-1
0-31/*: edu.stanford.nlp.ie.crf.CRFClassifier -loadClassifier
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/pku.gz -keepAllWhitespaces false -textFile input.txt
-serDictionary
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/dict-chris6.ser.gz -sighanCorporaDict
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/./data/ -sighanPostProcessing true -inputEncoding UTF-8

7. not work (replace absolute path to relative path from 4)

  /usr/bin/java -mx2g -cp
./stanford-segmenter-3.7.0.jar:./slf4j-1.7.25/slf4j-api-1.7.25.jar
edu.stanford.nlp.ie.crf.CRFClassifier -loadClassifier
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/pku.gz -keepAllWhitespaces false -textFile input.txt
-serDictionary
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/dict-chris6.ser.gz -sighanCorporaDict
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/./data/ -sighanPostProcessing true -inputEncoding UTF-8

8. work (remove slf4j from 7)

  /usr/bin/java -mx2g -cp ./stanford-segmenter-3.7.0.jar:
edu.stanford.nlp.ie.crf.CRFClassifier -loadClassifier
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/pku.gz -keepAllWhitespaces false -textFile input.txt
-serDictionary
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/dict-chris6.ser.gz -sighanCorporaDict
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/./data/ -sighanPostProcessing true -inputEncoding UTF-8

9. not work (remove segmenter from 7)

  /usr/bin/java -mx2g -cp ./slf4j-1.7.25/slf4j-api-1.7.25.jar:
edu.stanford.nlp.ie.crf.CRFClassifier -loadClassifier
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/pku.gz -keepAllWhitespaces false -textFile input.txt
-serDictionary
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/data/dict-chris6.ser.gz -sighanCorporaDict
/Users/xu2333/Documents/4Paradigm/stanfordNLP/stanford-segmenter-2016-10
-31/./data/ -sighanPostProcessing true -inputEncoding UTF-8

Now the reason is quite clear, the new version of
stanford-segmenter-2016-10-31 don't need the path of slf4j.
This modification could fix the bug of both Chinese segmenter and
Arabic segmenter.

xu2333 committed:
Add name

That’s exciting.

xu2333 commented:
That's exactly the issue #1652

stevenbird commented:
@dimazest would you please review

dimazest commented on nltk/tokenize/stanford_segmenter.py:
```diff
@@ -70,9 +70,8 @@ def __init__(self,
                 searchpath=(), url=_stanford_url,
```
if `slf4j` is not used, then the line 67 can be removed. Also, `self._SLF4J` is not needed.

dimazest replied:
`path_to_slf4j` also should be removed, line 52.

dimazest replied:
and line 49 is redundant.

xu2333 replied:
I could modify these! And I think we could write a version checker or something to help.

xu2333 replied:
I have removed the redundant parts and finished the tests locally. In addition, I think since the stanford-segmenter-2016-10-31 is released 7 months ago, it's acceptable to move forward.

dimazest commented on nltk/tokenize/stanford_segmenter.py:
```diff
@@ -70,9 +70,8 @@ def __init__(self,
-        self._stanford_jar = os.pathsep.join(
-            [_ for _ in [stanford_segmenter, slf4j] if not _ is None])
+        # Update: The new version of stanford-segmenter-2016-10-31 don't need the path of slf4j
+        self._stanford_jar = stanford_segmenter
```
should it be a list?

something like `self._stanford_jar = [stanford_segmenter]`

xu2333 replied:
No, the previous one is also a string because it's using "join", which is concatenating the two element.

dimazest reviewed (changes requested):
I would keep the current code, because an old version of stanford-segmenter might be used, then slf4 is needed.

dimazest commented:
[CI: retest]

xu2333 committed:
Remove the redundant parameters for slf4j

dimazest commented:
On the other side, one might need to be able to use an older version, for example, to replicate an experiment. I'm not sure whether this change is necessary.

xu2333 commented:
Actually, the update of stanford-segmenter.jar only need to modify path parameter point to the .jar file. 
And the stanfordNLP's Named Entity Recognizer of Chinese is relied on the output of Segmenter of Chinese. I read the recommendation for nltk from the homepage of stanfordNLP's NER and I think other students would do the same. I think this modification would benefit them a lot from those cmd bugs.

dimazest commented:
Do you mean that without this patch the code doesn't work?

dimazest commented on nltk/tokenize/stanford_segmenter.py:
```diff
@@ -46,10 +46,9 @@ class StanfordSegmenter(TokenizerI):
-    _SLF4J = 'slf4j-api.jar'
 
     def __init__(self,
-                 path_to_jar=None, path_to_slf4j=None,
```
keep `path_to_slf4j`, but make it default to `'slf4j-api.jar'`. Also move it to the new line.

dimazest commented on nltk/tokenize/stanford_segmenter.py:
```diff
@@ -64,15 +63,9 @@ def __init__(self,
                 env_vars=('STANFORD_SEGMENTER',),
                 searchpath=(), url=_stanford_url,
                 verbose=verbose)
-        slf4j = find_jar(
```
now here do something like
```python
if path_to_slf4j is not None:
    slf4j = find_jar(...)
else:
    slf4j = None

xu2333 replied:
I have finished the tests locally. The reason for using 'PLACEHOLDER' as default value is that  if path_to_slf4j is not None (like 'slf4j-api.jar'), the find_jar_iter in internals.py would execute:

    if path_to_jar is not None:
        if os.path.isfile(path_to_jar):
            yielded = True
            yield path_to_jar
        else:
            raise LookupError('Could not find %s jar file at %s' %
                            (name_pattern, path_to_jar))

But we should search the file using environment variables. Hence in default situation of using old version, we need to set path_to_slf4j to None.

dimazest replied:
but in my example `find_jar` will not be called if `path_to_slf4j` is `None`.

dimazest replied:
Did you try my proposal? Something like:
```python
StanfordSegmenter(path_to_slf4j=None)
```

you have set `path_to_slf4j` to force exclusion of the `slf4j`. This is explicit and compatible with the current behavior.

dimazest reviewed (changes requested):
Once these changes are applied, by setting `path_to_slf4j` to `None` should not append the slf4j.

dimazest commented:
By the way, have you considered using the CoreNLP instead? https://github.com/nltk/nltk/blob/develop/nltk/parse/corenlp.py

It should be faster, but the segmenter might need to be implemented.

alvations commented:
The segmenter might need some additional work to the interface `nltk.parse.corenlp` because it needs to use different properties for different languages when firing the corenlp server from java.

xu2333 committed:
changes according to dimazest's advice

I have read the code of find_jar_iter in the internals.py.
If path_to_slf4j is set to 'slf4j-api.jar' and input to find_jar()
directly, the searching process will execute:

    if path_to_jar is not None:
        if os.path.isfile(path_to_jar):
            yielded = True
            yield path_to_jar
        else:
            raise LookupError('Could not find %s jar file at %s' %
                            (name_pattern, path_to_jar))

and ignore the environment variables.  So in the default situation we
need to set path_to_jar to None.

dimazest commented:
[unrelated to this PR]

@alvations does the returned json depend on the language? We need to think how to support multiple languages with the CoreNLP interface.

dimazest commented on nltk/tokenize/stanford_segmenter.py:
```diff
@@ -64,13 +65,20 @@ def __init__(self,
                 verbose=verbose)
-        slf4j = find_jar(
+        if path_to_slf4j is not None:
+            if path_to_slf4j == 'PLACEHOLDER':
```
this will lead to a failure if ``path_to_slf4j is None`.

xu2333 replied:
The 'PLACEHOLDER' is to enable the old version to execute.
When using old version, and write

segmenter = StanfordSegmenter()

We need to find the correct slf4j-api.jar, and the find_jar_iter in internals.py would execute the following:

    if path_to_jar is not None:
        if os.path.isfile(path_to_jar):
            yielded = True
            yield path_to_jar
        else:
            raise LookupError('Could not find %s jar file at %s' %
                            (name_pattern, path_to_jar))

and ignore the environment variables. We can only find correct slf4j-api.jar using the environment variables. So in this default situation, we need to set path_to_jar, which is the second parameter transferred to find_jar(), to None.

xu2333 replied:
For new version, you are right, we need to write:

StanfordSegmenter(path_to_slf4j=None)

I have tested it locally and it works.

dimazest commented on nltk/tokenize/stanford_segmenter.py:
```diff
@@ -64,13 +65,20 @@ def __init__(self,
 
-        # This is passed to java as the -cp option, the segmenter needs slf4j.
+        # This is passed to java as the -cp option, the old version of segmenter needs slf4j.
+        # The new version of stanford-segmenter-2016-10-31 doesn't need slf4j
```
I would put this text to the docstring. Something like "if stanford-segmenter-2016-10-31 or newer is used, then path_to_slf4j should explicitly be set to None".

dimazest replied:
Unless, NLTK detects the segmenter's version and makes the decision.

stevenbird replied:
How about the converse of the first proposal, where using the more recent versions is the default. Users will get a runtime error when they upgrade NLTK, but the docstring (and the release notes) can cover for this.

xu2333 replied:
I think take the recent version as the default is workable. The question is how to set the parameter of old version. I am waiting for dimazest's opinion.

dimazest replied:
You can set the default value of `path_to_slf4j` to `None`. And then have an if like
```python
if path_to_slf4j is not None:
    slf4j = find_jar(path_to_slf4j=path_to_slf4j, ...)
else:
    slf4j = None
```

Then the only missing bit is the documentation, which needs to tell that if an old version if used, then a path to `path_to_slf4j` has to be set (most probably, to `'slf4j-api.jar'`.

xu2333 replied:
OK, I think the current code meets the requirement you mentioned.

dimazest replied:
Yes, it looks so. Could you please document this behavior and mention that if an old version of the segmenter is used, then the `path_to_slf4j` has to be set and a good value for it is `'slf4j-api.jar'`.

Also line 49 `_SLF4J = 'slf4j-api.jar'` needs to be cleaned up.

xu2333 replied:
OK. I have deleted _SLF4J. 
As I have recorded, as long as the second parameter transmitted to find_jar is not None, the find_jar_iter would execute without environment parameters. So If the user only input 'slf4j-api.jar' instead of the path to slf4j-api.jar, I need to set this parameter to None.

dimazest replied:
does it require a full path, i guess, a relative path (just 'slf4j-api.jar') should also work.

xu2333 committed:
enable new version to execute without parameter

finished the update according to stevenbird and dimazest's advice and finished tests locally.

xu2333 committed:
delete line 49 and modify the explanation

xu2333 committed:
update the parameter for find_jar

Prevent the find_jar_iter from terminating before searching using environment parameters

dimazest commented on nltk/tokenize/stanford_segmenter.py:
```diff
@@ -76,7 +75,7 @@ def __init__(self,
         # This is passed to java as the -cp option, the old version of segmenter needs slf4j.
         # The new version of stanford-segmenter-2016-10-31 doesn't need slf4j
-        # When using old version, set 'path_to_slf4j=YOUR_PATH' in the initilization of StanfordSegmenter
+        # When using old version, set path_to_slf4j='/YOUR_PATH/slf4j-api.jar' in the initilization of StanfordSegmenter
```
Could you please move this note into the docstring that starts on the line 33. It's also better to specify concretely that versions before stanford-segmenter-2016-10-31 need to be used like this.

dimazest replied:
Something like
```
If stanford-segmenter version is older than 2016-10-31, then path_to_slf4j should explicitly be set to '/YOUR_PATH/slf4j-api.jar'::

   segmenter = StanfordSegmenter(path_to_slf4j='/YOUR_PATH/slf4j-api.jar')
```

xu2333 replied:
I think you have decided that we would choose the plan A, which is letting user input the full path of slf4j-api.jar. Hence I change the parameter back to path_to_slf4j.

xu2333 committed:
update for docstring

dimazest commented on nltk/tokenize/stanford_segmenter.py:
```diff
@@ -31,6 +31,8 @@
 class StanfordSegmenter(TokenizerI):
     """
     Interface to the Stanford Segmenter
```
The final minor thing, could you format this nicely:
```python
"""Interface to the Stanford Segmenter.

If stanford-segmenter version is older than 2016-10-31, then path_to_slf4j
should be provieded, for example::

    seg = StanfordSegmenter(path_to_slf4j='/YOUR_PATH/slf4j-api.jar')

>>> from nltk.tokenize.stanford_segmenter import StanfordSegmenter
>>> seg = StanfordSegmenter()
>>> seg.default_config('zh')
...

"""
```

dimazest commented on nltk/tokenize/stanford_segmenter.py:
```diff
@@ -75,8 +77,6 @@ def __init__(self,
-        # When using old version, set path_to_slf4j='slf4j-api.jar' in the initilization of StanfordSegmenter
-        # The initilization program will search for your slf4j-api.jar according to the environment parameters you set
         self._stanford_jar = os.pathsep.join(
             [_ for _ in [stanford_segmenter, slf4j] if not _ is None])
```
I'm not sure that a list needs to be built here, try
```python
self._stanford_jar = os.pathsep.join(
    _ for _ in [stanford_segmenter, slf4j] if _ is not None
)
```

Also `if not _ is None` is strange `_ is not None` is much better.

xu2333 replied:
Yes, you are right, join() could take a sequence as input. And of course, _ is not None is more natural.

xu2333 replied:
And I am sure that a relative path would not work. When you execute find_jar(), whose second parameter is path_to_slf4j, it would directly search according to the path without environment parameters and then terminate. So we have 3 choices:
A. let user input full path
B. set the second parameter of find_jar() to None when path_to_slf4j is not None, which means that the path user input is just a signal to make sure that it's old version
C. revise the find_jar_iter() function in internals.py

xu2333 committed:
update according to dimazest's advice

dimazest commented:
👍 Looks good.

stevenbird commented:
@xu2333: thanks for your contribution
@dimazest: thanks for your review. Can you confirm whether [this comment of yours](https://github.com/nltk/nltk/pull/1743#issuecomment-305541086) is still relevant?

xu2333 committed:
Merge branch 'develop' into hotfix/bug-on-stanford-segmenter

dimazest commented:
@stevenbird that was addressed:  it's still possible to use an older version of Stanford tools.

stevenbird commented:
Thanks @dimazest!
